### PR TITLE
[Feature]: Refactor to Use The API And Styling

### PR DESCRIPTION
**Pull request by mohamedSabry0**
in this PR:
- Using Bookstore API as the app's backend.
- Refactoring redux reducers to use [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) and asynchronous thunks.
- Working on the styling to match the provided design as much as possible.

### CODE REVIEW

**Comment by mohamedSabry0** [comment]
**A note for the reviewer of [styling milestone]: Sorry, I did submit the URL in the wrong place by mistake.**
you can go ahead and invalidate it, or as you see fit.

File: src/components/Books.js [hunk 1]
```diff
@@ -1,22 +1,10 @@
-import { useSelector } from 'react-redux';
```
Comment (Kamba56) [inline src/components/Books.js:1]: - [ ] Kindly style your project to look identical to this template as demanded by the project requirements. Kindly ensure that all components on the template are added as placeholders without any added functionality and also ensure that the functionalities of the bookstore are retained.
image

File: src/redux/books/booksSlice.js [hunk 2]
```diff
@@ -1,43 +1,44 @@
...
+      .addCase(addBook.fulfilled, (state, action) => ({
+        ...state,
+        books: [...state.books, action.payload],
+        status: 'succeeded',
+      }))
+      .addCase(removeBook.fulfilled, (state, action) => ({
+        ...state,
+        books: [...state.books.filter((item) => item.item_id !== action.payload)],
+        status: 'succeeded',
+      }));
```
Comment (Kamba56) [inline src/redux/books/booksSlice.js]: - [Optional] Currently, when a user adds or removes a book, the change is only visible after the user refreshes the page. Kindly consider making the change visible immediately after a book has been added or removed by updating the redux store after the book has been successfully added/removed.

**Review by Kamba56 — CHANGES_REQUESTED** [review]
Hi @mohamedSabry0, 

Good job so far!!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏!👏👏👏
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- [x] Great use of Github workflow!👏👏👏
- [x] All linters passed

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/redux/books/booksThunks.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        throw new Error(`HTTP error! Status: ${res.status}`);
+      }
+      return res;
+    })
+    .catch((err) => err.message);
+  if (response.statusText !== 'Created') {
+    return 'failed';
+  }
+  return id;
+});
```
Comment (franclobo) [inline src/redux/books/booksThunks.js]: - [ ] There is an issue in the add and remove functions that take action after refreshing the page, please 🙏🏽 correct this behavior for a better UX.

>  When a user adds or removes a book, the change is only visible after the user refreshes the page. Kindly consider making the change visible immediately after a book has been added or removed by updating the redux store after the book has been successfully added/removed.

**Review by franclobo — CHANGES_REQUESTED** [review]
Hi @mohamedSabry0, 

## ♻️ Required Changes ♻️

so-close-jim-halpert

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights

- `fetch` is used to make the API calls ✔️
- `createAsyncThunk` is used ✔️ 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/redux/books/booksSlice.js [hunk 4]
```diff
@@ -1,43 +1,56 @@
...
+        error: action.error.message,
+        status: 'failed',
+      }));
   },
 });
-
-export const { removeBook, addBook } = booksSlice.actions;
+export const booksState = (state) => state.books;
+export { fetchBooks, addBook, removeBook };
+// export const { removeBook } = booksSlice.actions;
```
Comment (tresorsawasawa) [inline src/redux/books/booksSlice.js]: - [ OPTIONAL ] Kindly make sure you remove all commented and unused code lines from your code base to make it cleaner, professional and for best practices

File: src/redux/books/booksThunks.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    .then((res) => {
+      if (!res.ok) {
+        throw new Error(`HTTP error! Status: ${res.status}`);
+      }
+      return res;
+    });
+  if (response.statusText !== 'Created') {
+    return response;
+  }
+  return id;
```
Comment (tresorsawasawa) [inline src/redux/books/booksThunks.js]: - [ ] As required for the project everything to still work as expected and as mentioned by the previous reviewers,  there is still an issue when a use try to `add` and `remove` a book. The action takes after refreshing the page, please 🙏🏽 correct this behavior for a better UX.

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUEST CHANGES :recycle::recycle: 

Hi @mohamedSabry0 
Great work on making the changes requested by the previous reviewer  👏👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost-there

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by rotimiazeez — COMMENTED** [review]
# Hi @mohamedSabry0, 

### While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that fails some linter test.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit._

**Review by rotimiazeez — COMMENTED** [review]
# Hi @mohamedSabry0 , 

### While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project that needs to pass the linter checks. Kindly fix this issue before submitting again 👍🏽 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit._

**Review by victorokoroji — APPROVED** [review]
Hi @mohamedSabry0, 

# ✔️   Status: APPROVED 🏆 🎉 🎊 

You have done a great job on this project, keep it up 👍 🥇. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
congratulations!  🎉💯🌟 

## Top highlight 🎯

- The Pull Request is following the GitHub Flow ✅ 
- The PR title and description are professional and descriptive ✅ 
- Linters are working properly ✅ 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Deepakdanger — APPROVED** [review]
Hi @mohamedSabry0, 

### Status: Approved ✔️

Your project is complete!  There is nothing else to say other than... it's time to merge it  🚢 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**